libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -673,32 +566,44 @@ void DeclarationRegistrationHelper::endVisit(VariableDeclarationStatement& _vari
 	for (ASTPointer<VariableDeclaration> const& var: _variableDeclarationStatement.declarations())
 		if (var)
 			m_currentFunction->addLocalVariable(*var);
+	ASTVisitor::endVisit(_variableDeclarationStatement);
 }
 
-bool DeclarationRegistrationHelper::visit(VariableDeclaration& _declaration)
+bool DeclarationRegistrationHelper::visitNode(ASTNode& _node)
 {
-	registerDeclaration(_declaration, false);
-	return true;
-}
+	if (auto const* scopable = dynamic_cast<Scopable const*>(&_node))
+		solAssert(scopable->annotation().scope == m_currentScope, "");
+
+	if (auto* declaration = dynamic_cast<Declaration*>(&_node))
+		registerDeclaration(*declaration);
+	if (dynamic_cast<ScopeOpener const*>(&_node))
+		enterNewSubScope(_node);
+
+	if (dynamic_cast<ScopeOpener const*>(&_node))
+		m_currentScope = &_node;
+	if (auto* variableScope = dynamic_cast<VariableScope*>(&_node))
+		m_currentFunction = variableScope;
+	if (auto* annotation = dynamic_cast<TypeDeclarationAnnotation*>(&_node.annotation()))
+		annotation->canonicalName = currentCanonicalName();
 
-bool DeclarationRegistrationHelper::visit(EventDefinition& _event)
-{
-	registerDeclaration(_event, true);
 	return true;
 }
 
-void DeclarationRegistrationHelper::endVisit(EventDefinition&)
+void DeclarationRegistrationHelper::endVisitNode(ASTNode& _node)
 {
-	closeCurrentScope();
+	if (dynamic_cast<ScopeOpener const*>(&_node))
+		closeCurrentScope();
+	if (dynamic_cast<VariableScope*>(&_node))
+		m_currentFunction = nullptr;
 }
 
 void DeclarationRegistrationHelper::enterNewSubScope(ASTNode& _subScope)
 {
-	map<ASTNode const*, shared_ptr<DeclarationContainer>>::iterator iter;
-	bool newlyAdded;
 	shared_ptr<DeclarationContainer> container{make_shared<DeclarationContainer>(m_currentScope, m_scopes[m_currentScope].get())};
-	tie(iter, newlyAdded) = m_scopes.emplace(&_subScope, move(container));
-	solAssert(newlyAdded, "Unable to add new scope.");
+	bool newlyAdded = m_scopes.emplace(&_subScope, move(container)).second;
+	// Source units are the only AST nodes for which containers can be created from multiple places
+	// due to imports.
+	solAssert(newlyAdded || dynamic_cast<SourceUnit const*>(&_subScope), "Unable to add new scope.");
```

Comment:
What would be an example where `newlyAdded` would be false?

Reply:
In an import directive, the scope of the imported file might already be created: https://github.com/ethereum/solidity/pull/9586/files#diff-653a1e23aa75715b13289b03e2f20947R536